### DISCUSSION TITLE
Applying schema snapshot has no effect

### DISCUSSION THREAD

**Discussion by nhh** [directus/directus#16407] [Archive]
### Describe the Bug

Referring to your docs, you are mentioning that I can use schema snapshots for environment changes. I am evaulating this for our local development setup. (1 dev changes the schema, another is applying the snapshot)

I followed the docs and ran the command "npx directus schema apply" inside the directus docker container. Unfortunately nothing happens. 

It is a fresh database. I expected to see my collections inside directus.

### To Reproduce

1. Setup directus locally. (inside docker)
2. Add collections and your schema
3. Create a schema snapshot inside container (and copy it out)
4. Reset database
5. Rerun directus (applying initial migrations)
6. Apply schema snapshot with "npx directus schema apply"

### What version of Directus are you using?

9.20.4

### What version of Node.js are you using?

docker

### What database are you using?

Postgres 13

### What browser are you using?

Chrome

### How are you deploying Directus?

docker

**Comment by nhh** [comment]
Addition: The cli seems to be exiting with code 1 but all output is swallowed away.

`LOG_LEVEL=trace npx directus schema apply --yes /directus/schema.yml` 

Increasing the log level has no effect either.

**Comment by br41nslug** [comment]
The missing logs is most likely a consequence of this issue https://github.com/directus/directus/issues/16284

**Comment by azrikahar** [comment]
@nhh can you try setting `LOG_STYLE="raw"` to share what are the errors? (there's likely some kind of error rather than "no effect" due to the log issue as @br41nslug mentioned, but that should be temporarily circumvented with `LOG_STYLE="raw"`)

**Comment by nhh** [comment]
Wild. It seems to work now! I tested it by deleting columns and tables and applying the snapshot.

```
{"level":30,"time":1667997255032,"pid":450,"hostname":"d69e7f27cc82","msg":"Snapshot applied successfully"}
```

I could swear yestderday this did not work. ¯\_(ツ)_/¯

Thank you all! <3

**Reply by mfeitoza** [reply]
How? I trying many ways, but nothing works...

**Comment by nhh** [comment]
Are schema snapshots automatically create new collections? If not, this would explain why things didnt work. 😄 

So the question in question is: Is there a way to commit/versioning my database schema including the creation and deletion of a collection?

I would love to have migration tooling available. Otherwise I would have to create every collection per hand for every environment (dev, staging & prod).

Can you give me some advice? 😊

**Reply by azrikahar** [reply]
> Are schema snapshots automatically create new collections?

It will, provided the collection doesn't already exist on the target instance 👍

Schema snapshots are compared with the current instance's schema, and the diffs will be the resulting changes, if any.

I do recommend using the [`--dry-run` flag](https://docs.directus.io/reference/cli.html#applying-a-snapshot) to review the changes of the schema apply before proceeding with the actual apply (since you used `--yes`, you might not know what was the actual diffs).

> So the question in question is: Is there a way to commit/versioning my database schema including the creation and deletion of a collection?

The schema snapshot technically encapsulates the schema at the time of snapshot, rather than a "timeline" or migrations, if that makes sense.

There's also [custom migrations](https://docs.directus.io/extensions/migrations.html) if there are any migrations you are looking to perform but not included in schema migrations (such as data).

**Reply by nhh** [reply]
This is something I need to dig down deeper. Thanks for the help so far :)

**Comment by JulianBissekkou** [comment]
I was able to reproduce this using directus 9.21.0.
Applying a schema on a clean and empty directus instance using `npx directus schema apply schema.yaml --yes` does nothing.

Executing the statement with `LOG_STYLE="raw"` prints the error that happens:
```
{"level":50,"time":1669119031472,"pid":256,"hostname":"ff81210b81d3","msg":"Failed to create relation \"block.audio_file\""}
{"level":50,"time":1669119031475,"pid":256,"hostname":"ff81210b81d3","err":{"type":"InvalidPayloadException","message":"Collection \"block\" doesn't exist","stack":"Error: Collection \"block\" doesn't exist\n    at RelationsService.createOne (/directus/node_modules/directus/dist/services/relations.js:141:19)\n    at /directus/node_modules/directus/dist/utils/apply-snapshot.js:190:44","status":400,"code":"INVALID_PAYLOAD","extensions":{}},"msg":"Collection \"block\" doesn't exist"}
```

However I can see that the collection is created when running the command with `LOG_LEVEL="trace"`:

```
...
[12:11:31] TRACE (278 on ff81210b81d3): [0.455ms] select "collection" from "directus_collections" []
[12:11:31] TRACE (278 on ff81210b81d3): [4.106ms] create table "block" ("audio_file" uuid null default null, "audio_file_phrase_shower" uuid null default null, "date_created" timestamptz null default null, "date_updated" timestamptz null default null, "id" serial primary key, "language_pair" json null default null, "section" integer null default null, "sort" integer null default null, "user_created" uuid null default null, "user_updated" uuid null default null) []
...
```

I can share the schema in a private chat on discord if you like.

**Reply by azrikahar** [reply]
Should be the same issue as #16503 which is present in `9.21.0` 👍

**Reply by JulianBissekkou** [reply]
I see that there was a PR merged. When can we expect a release of that bugfix?

**Reply by azrikahar** [reply]
Unfortunately as of now I'm not personally aware of a specific date/time for the next release as it can be subject to change.

**Comment by rupampoddar** [comment]
This worked for me:

on local system:
- downgrade directus to `9.20.4` 
- revert db migration `npx directus database migrate:down`  
- create new snapshot
- upload the snapshot file to server

on server (fresh server, fresh db):
- npx directus bootstrap
- apply the schema snapshot
